torchtune/datasets/_chat.py

```diff
@@ -81,65 +76,23 @@ def __getitem__(self, index: int) -> Tuple[List[int], List[int]]:
         return self._prepare_sample(sample)
 
     def _prepare_sample(self, sample: Mapping[str, Any]) -> Tuple[List[int], List[int]]:
-        dialogue = self._convert_to_dialogue(sample)
-
-        prompt_tokens = []
-        label_tokens = []
-        for prompt, label in self._get_turns(dialogue):
-            formatted_prompt = self.template.format(prompt)
-            encoded_prompt_with_response, labels = tokenize_prompt_and_response(
-                tokenizer=self._tokenizer,
-                prompt=formatted_prompt,
-                response=label,
-                train_on_input=self.train_on_input,
-            )
-            prompt_tokens.extend(encoded_prompt_with_response)
-            label_tokens.extend(labels)
-
-            if len(prompt_tokens) >= self.max_seq_len:
-                break
-
-        prompt_tokens, label_tokens = truncate(
-            self._tokenizer, prompt_tokens, label_tokens, self.max_seq_len
+        messages = self._convert_to_messages(sample)
+        messages = self.chat_format.format(messages)
+        tokens, mask = self._tokenizer.tokenize_messages(
+            messages, max_seq_len=self.max_seq_len
         )
+        labels = list(np.where(np.logical_not(mask), tokens, CROSS_ENTROPY_IGNORE_IDX))
```

Comment:
Very nice

Reply:
Nice indeed, but please add comment for mere mortals like me who need to take more than a min to understand this :)

Reply:
I made this needlessly complicated so that's my fault